internal/impl/tigerbeetle/input_tigerbeetle.go

```diff
@@ -0,0 +1,646 @@
+}
+
+func (input *tigerbeetleInput) ReadBatch(ctx context.Context) (service.MessageBatch, service.AckFunc, error) {
+	select {
+	case batchedMessage := <-input.consumerChan:
+		return batchedMessage.batch, batchedMessage.ackFunc, nil
+	case <-input.stopSignaller.HasStoppedChan():
+		return nil, nil, service.ErrNotConnected
+	case <-ctx.Done():
+	}
+	return nil, nil, ctx.Err()
+}
+
+func newTigerbeetleInput(config *service.ParsedConfig, resources *service.Resources) (s service.BatchInput, err error) {
+	var (
+		clusterID           string
+		addresses           []string
+		eventCountMax       int
+		idleInterval        int
+		timestampInitialStr string
+		timestampInitial    uint64 = 0
+		progressCache       string
+	)
+
+	logger := resources.Logger()
+
+	if clusterID, err = config.FieldString(fieldClusterID); err != nil {
+		return nil, err
+	}
+	clusterID128, success := StringToUint128(clusterID)
+	if !success {
+		return nil, fmt.Errorf("invalid config: %s='%s'", fieldClusterID, clusterID)
+	}
+
+	if addresses, err = config.FieldStringList(fieldAddresses); err != nil {
+		return nil, err
+	}
+	if len(addresses) == 0 {
+		return nil, fmt.Errorf("invalid config: %s is empty", fieldAddresses)
+	}
+
+	if progressCache, err = config.FieldString(fieldProgressCache); err != nil {
+		return nil, err
+	}
+	if !config.Resources().HasCache(progressCache) {
+		return nil, fmt.Errorf("cache resource '%s' not found", progressCache)
+	}
+
+	if eventCountMax, err = config.FieldInt(fieldEventCountMax); err != nil {
+		return nil, err
+	} else if eventCountMax <= 0 {
+		return nil, fmt.Errorf("property '%s' must be greater than zero", fieldEventCountMax)
+	}
+
+	if idleInterval, err = config.FieldInt(fieldIdleInterval); err != nil {
+		return nil, err
+	} else if idleInterval <= 0 {
+		return nil, fmt.Errorf("property '%s' must be greater than zero", fieldIdleInterval)
+	}
+
+	if timestampInitialStr, err = config.FieldString(fieldTimestampInitial); err != nil {
+		return nil, err
+	} else if len(timestampInitialStr) != 0 {
+		if timestampInitial, err = strconv.ParseUint(timestampInitialStr, 10, 64); err != nil {
+			return nil, fmt.Errorf("invalid config: %s='%s'", fieldTimestampInitial, timestampInitialStr)
+		}
+	}
+
+	input := &tigerbeetleInput{
+		config: tigerbeetleConfig{
+			clusterID:        clusterID128,
+			addresses:        addresses,
+			progressCache:    progressCache,
+			timestampLastKey: "timestamp_last_" + clusterID,
+			eventCountMax:    uint32(eventCountMax),
+			idleInterval:     time.Duration(idleInterval) * time.Millisecond,
+			timestampInitial: timestampInitial,
+		},
+		producerChan: make(chan []tb_types.ChangeEvent, 1),
+		consumerChan: make(chan batchedMesssage, 1),
+		logger:       logger,
+		resources:    resources,
+	}
+
+	return service.AutoRetryNacksBatchedToggled(config, input)
+}
+
+// Extracts events from TigerBeetle.
+func (input *tigerbeetleInput) produce(ctx context.Context, client tb.Client, timestampLast uint64) error {
```

Comment:
nit: how about using short variable name for self?

Reply:
I guess it would usually be named just `i` in Go!

I ended up using the [TigerStyle](https://github.com/tigerbeetle/tigerbeetle/blob/main/docs/TIGER_STYLE.md#naming-things) coding style, favoring short but meaningful names and avoiding single-letter variables and abbreviations (except for trivial counters/indexes and well-known abbreviations like `ctx`, `config`, etc).  

Although TigerStyle was originally designed for another programming language (Zig), I think we can apply most of the rules as long as they don't conflict with Go's idiomatic conventions. I don't want this code to feel alien 👽 compared with the rest of the code base!

It's your call, I'd be glad to follow your coding style otherwise.